### ISSUE TITLE
Ctrl collection highlighting causing problems on Linux (and Windows?)

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#3782]
https://forums.zotero.org/discussion/112327/zotero-7-beta-conflict-with-the-ctrl-key-on-linux

Interfering with Ctrl+PgDn/PgUp is definitely a problem, but I'm confused about the other. Has this always been happening on Windows (where Ctrl has been used for collection highlighting)? Has it been highlighting and possibly scrolling the collections pane any time you used a shortcut key? For some reason collection highlighting isn't working for me at all on Windows, so I can't test this.

**Comment by abaevbog** [comment]
> Interfering with Ctrl+PgDn/PgUp is definitely a problem

Looks like this is an edge case that I missed while testing if there are any issues with making the `ZoteroPane.handleKeyDown` a non-capturing listener to avoid all sorts of random issues with focus https://github.com/zotero/zotero/commit/5f0e3d550bf4fabefa8d6a1392d24018021c7f69. It looks like  Ctrl+PgDn/PgUp must be capturing because if the focus is on the itemTree in the library tab, it ends up being picked up by the scrollable area of itemTree and never reaches the handler that changes the selected tab. 


> Has it been highlighting and possibly scrolling the collections pane any time you used a shortcut key?

This is a regression from 527fa5b12f33e7b79e77e65ad12c497d293768c9 - I misunderstood the purpose of the removed conditional at the bottom (partly because of the `// Unhighlight on key up` comment). That part is actually importabt because it's responsible for stopping the highlighting from kicking in if something else was pressed.

**Comment by abaevbog** [comment]
So we certainly want to bring back cancelation of collection highlighting if any other control character is pressed (just with a better comment), and I thought we can just have 2 `keydown` listeners: a bubbling and a capturing one? Sometimes we want to make sure the events are not handled by any other components (Ctrl+PgDn/PgUp) but as a general rule, we want to only handle events if no other listener took care of them (e.g. components should have the option to handle Escape if needed when reader is opened).

**Comment by adomasven** [comment]
@abaevbog I haven't looked at the details of what's causing issues here, but it seems that you may be somewhat confused about JS event handlers.

A bubbling and capturing event handler differs only in when in the event bubbling phase the event is handled. All JS events start at the currently active node and bubble all the way to the top node, then cascade back all the way back to the source node. Normal event handlers handle the event on the bubbling way to the top, capturing event handlers handle the event on the way back. If an event is handled by multiple handlers on nodes at different levels, any handler that is invoked earlier can call `event.stopPropagation()` to stop the event from bubbling (or cascading) further. There's also `event.stopImmediatePropagation()` which prevents the event from bubbling and firing on other event handlers registered for the same node. And `event.preventDefault()`, which e.g. prevents an `<input>` node from having its value changed by default browser handlers.

I hope this helps.

**Comment by abaevbog** [comment]
@adomasven, thank you for the great explanation! Intuitively, I think I had the somewhat correct idea but I may have done a terrible job with the explanation. Let me try again with some more details.

- The issue with collection highlighting being activated and scrolling when you're trying to use a shortcut with `Ctrl` (like `Ctrl-Shift-A`) just boils down to a regression after https://github.com/zotero/zotero/commit/527fa5b12f33e7b79e77e65ad12c497d293768c9 where the collection highlighting should stop if something but `Ctrl` is being pressed. That is addressed in the linked PR and has nothing to do with capturing/bubbling listeners.
- There were a few issues caused by the events firing at `ZoteroPane.handleKeyDown` before event listeners of lower level components and conflicting with them prior to https://github.com/zotero/zotero/commit/5f0e3d550bf4fabefa8d6a1392d24018021c7f69. This is a recent [example](https://github.com/zotero/zotero/issues/3725#issuecomment-1957855173). Another example was `Escape` keypress from context pane would always fire at `ZoteroPane.handleKeyDown` and focus the reader, even if it was meant to just close a context menu (e.g. `itemType` dropdown from `itemBox` in contextPane). After https://github.com/zotero/zotero/commit/5f0e3d550bf4fabefa8d6a1392d24018021c7f69, `ZoteroPane.handleKeyDown` stopped capturing events to make sure nothing runs before lower level components get a chance to handle events, which addressed the issues above. I did miss that there are cases in which a lower-level component may `stopPropagation()` of an event, which will mean that it never bubbles to `ZoteroPane.handleKeyDown`. This is what happens with Ctrl+PgDn/PgUp when itemTree is focused. So as a solution, I am suggesting to have 2 `keypress` handlers. The capturing one is for events related to tab navigation shortcuts, since it runs before any lower-level component stops the event's propagation (this fixes the current Ctrl+PgDn/PgUp issue for me). The non-capturing listener will run only if the event bubbled all the way up and wasn't stopped by any other component which we is good for the `Escape` handling to not be triggered on popup closing from contextPane, since such event's propagation would be stopped.

Does this make more sense? Correct me if any of this is off!

**Comment by adomasven** [comment]
> * There were a few issues caused by the events firing at `ZoteroPane.handleKeyDown` before event listeners of lower level components and conflicting with them prior to [5f0e3d5](https://github.com/zotero/zotero/commit/5f0e3d550bf4fabefa8d6a1392d24018021c7f69). This is a recent [example](https://github.com/zotero/zotero/issues/3725#issuecomment-1957855173). Another example was `Escape` keypress from context pane would always fire at `ZoteroPane.handleKeyDown` and focus the reader, even if it was meant to just close a context menu (e.g. `itemType` dropdown from `itemBox` in contextPane). After [5f0e3d5](https://github.com/zotero/zotero/commit/5f0e3d550bf4fabefa8d6a1392d24018021c7f69), `ZoteroPane.handleKeyDown` stopped capturing events to make sure nothing runs before lower level components get a chance to handle events, which addressed the issues above. I did miss that there are cases in which a lower-level component may `stopPropagation()` of an event, which will mean that it never bubbles to `ZoteroPane.handleKeyDown`. This is what happens with Ctrl+PgDn/PgUp when itemTree is focused. So as a solution, I am suggesting to have 2 `keypress` handlers. The capturing one is for events related to tab navigation shortcuts, since it runs before any lower-level component stops the event's propagation (this fixes the current Ctrl+PgDn/PgUp issue for me). The non-capturing listener will run only if the event bubbled all the way up and wasn't stopped by any other component which we is good for the `Escape` handling to not be triggered on popup closing from contextPane, since such event's propagation would be stopped.

I did not bother to entirely follow all the changes and implications here, but it seems like we have an overly complex situation here. Ideally you only have bubbling or capturing event handlers (bubbling usually makes more sense), and any time you need to mix both, especially on the same element means there are likely design problems.

Some simple thinking here: If a lower-level component event handler handles a key for navigation, generally it means that a higher-level component shouldn't. In that case, the lower-level event handler should call `event.stopBubbling()`. Are we doing that? If not, why not? Would some of the issues here be solved if we did that? What wouldn't be solved?

**Comment by abaevbog** [comment]
> Ideally you only have bubbling or capturing event handlers (bubbling usually makes more sense)

I fully agree with that, having two listeners is indeed quite convoluted...

>  If a lower-level component event handler handles a key for navigation, generally it means that a higher-level component shouldn't

Our issue is that we want `ZoteroPane.handlekeyDown` to handle the event whose bubbling `itemTree` stops with `event.stopPropagation()` . It's [this code](https://github.com/zotero/zotero/blob/1ca23fedb2c81f713085c9f64e9ac22d2f88b96d/chrome/content/zotero/components/virtualized-table.jsx#L463) from `virtualized-table.jsx`. So this is actually the case for `collectionTree` too, not just `itemTree`.

So in fact, maybe a more correct solution would be to keep `ZoteroPane.handlekeyDown` as a single bubbling event handler and adjust `virtualized-table` so that the event's propagation is not stopped on Ctrl+PgDn/PgUp? I tried a small change from [this commit](https://github.com/abaevbog/zotero/commit/fe6a2d114606ffdc13b563e47916e00931f95216) to just not do anything on Ctrl+PgDn/PgUp and this seems to solve the issue as well without having a bubbling and capturing listener in `ZoteroPane`. What do you think about this? Since this shortcut is meant for tabs navigation and prior to https://github.com/zotero/zotero/commit/5f0e3d550bf4fabefa8d6a1392d24018021c7f69 it wasn't even reaching `virtualized-table` at all, this should not create any conflicts with existing keyboard navigation.

**Comment by adomasven** [comment]
> I tried a small change from [this commit](https://github.com/abaevbog/zotero/commit/fe6a2d114606ffdc13b563e47916e00931f95216) to just not do anything on Ctrl+PgDn/PgUp and this seems to solve the issue as well without having a bubbling and capturing listener in `ZoteroPane`.

Ok, so given the problem is that the virtualized-table is calling `stopPropagation()`, the best solution is not calling it there. I wrote that code, because without it using arrow keys (and page up/down on non-macOS) the browser performs a scroll on a scrollable element - the virtualized table. But that's stopped by calling `preventDefault()`. I'm not sure `stopPropagation()` does anything useful there. Can we not remove that without consequence?

**Comment by abaevbog** [comment]
After removing `event.stopPropagation` I did not see any changes in behavior with keyboard navigation on any of the platforms. So that's great!

I think we still want to keep the piece that explicitly stops handling of Ctrl+PageUp/PageDown though. Otherwise, the event will bubble to ZoteroPane and the tabs will be switched but `_onJumpSelect`/`jsWindow.scrollTo` from the `virtualized-table` will still run. So if `itemTree` is focused and you press  Ctrl+PageUp to switch tabs and Ctrl+PageDown to go back to the library tab, the itemTree will be in a different scroll position than before, and I don't think we want that? If this piece remains, removing `event.stopPropagation` is not technically required to solve this specific issue because `_preventKeyboardScrolling` is never reached but I suppose it's still worth it to avoid things like that in the future.

This is the change I was trying out https://github.com/zotero/zotero/pull/3793/commits/3372b7dd8f93340179c2674f9bc94c08207254dd

**Comment by adomasven** [comment]
Ok, so we seem to have conflicting hotkeys here. I didn't think about it, but I guess Ctrl-PgUp/Down actually moves the focused row by a full page in the virtualized table, which is technically useful as a behaviour, but I guess we want to forego that in favour of tab navigation in the main window. However, in other views where itemTable is used (and that will expand), Ctrl-PgUp/Down should still be handled by the virtualized table.

I looked up the event bubbling/capturing thing again, and realized that I had the logic backwards. The bubble first has a capturing phase, then a bubbling phase. So capturing event handlers allow you to implement stuff like global hotkeys. And now I reread your responses above and see the ZoteroPane event handler was capturing, then was switched to be non-capturing, but the best way to handle tab switching indeed just seems like a capturing event handler, so I guess your proposal of a separate capturing event handler for tab switching is indeed the best solution here. Sorry for making you meander.

**Comment by abaevbog** [comment]
> However, in other views where itemTable is used (and that will expand), Ctrl-PgUp/Down should still be handled by the virtualized table.

Oh yeah, I didn't think about virtualized tables that are not just `itemTree` or `collectionTree`.

>  so I guess your proposal of a separate capturing event handler for tab switching is indeed the best solution here. Sorry for making you meander.

No worries! I re-pushed that old commit back. It really does feel somewhat off that there are two top-level keydown listeners but ... I guess that's what it has to be
